src/components/common/ExpressionInput.vue

```diff
@@ -0,0 +1,170 @@
+    },
+    placeholder: {
+      type: String,
+      default: '',
+    },
+    currentExpression: {
+      type: String,
+      required: true,
+    },
+    expressionIdentifier: {
+      type: [String, Number],
+      default: null,
+    },
+  },
+
+  data() {
+    return {
+      suggest: {},
+      debounceScroll: () => undefined,
+    }
+  },
+
+  computed: {
+    ...mapGetters(['subscriberPropertyFields']),
+    ...mapGetters('flow', ['activeFlow']),
+
+    expression: {
+      get() {
+        return this.currentExpression
+      },
+      set(value) {
+        if (this.expressionIdentifier !== null) {
+          value = {
+            identifier: this.expressionIdentifier,
+            value,
+          }
+        }
+        this.$emit('commitExpressionChange', value)
+      },
+    },
+    expressionContext() {
+      const contactFields = this.subscriberPropertyFields.map(prop => prop.name).concat(defaultContactPropertyFields)
+      const blocks = this.activeFlow.blocks.flatMap(b => [b.name, b.uuid])
+      return {
+        contact: contactFields,
+        flow: blocks,
+        date: defaultDateFields,
+      }
+    },
+    topLevelSuggestions() {
+      return {
+        trigger: '@',
+        values: [
+          {
+            value: '@()',
+            focusText: [-1, -1],
+          },
+          ...Array.from(Object.entries(this.expressionContext)).map(item => `@${item[0]}`),
+        ],
+      }
+    },
+    contextSuggestions() {
+      return Array.from(Object.entries(this.expressionContext)).map((item) => {
+        const name = item[0]
+        return {
+          trigger: `${name}.`,
+          values: item[1].map(val => `${name}.${val}`),
+        }
+      })
+    },
+    methodSuggestions() {
+      return Array.from(this.evaluatorMethods.entries()).map(item => ({
+        trigger: item[0],
+        values: item[1].map(i => ({
+          value: `${i}()`,
+          focusText: [-1, -1],
+        })),
+      }))
+    },
+    evaluatorMethods() {
+      const methods = new Map()
+
+      /* eslint-disable no-restricted-syntax */
+      for (const handler of MethodNodeEvaluatorFactory.defaultHandlers()) {
+        for (const method of handler.handles()) {
+          const trigger = method.substr(0, 2)
+          const upperTrigger = trigger.toUpperCase()
+          const upperMethod = method.toUpperCase()
+          if (methods.has(trigger)) {
+            methods.set(trigger, [...methods.get(trigger), method])
+            methods.set(upperTrigger, [...methods.get(upperTrigger), upperMethod])
+          } else {
+            methods.set(trigger, [method])
+            methods.set(upperTrigger, [upperMethod])
+          }
+        }
+      }
+      /* eslint-enable no-restricted-syntax */
+      return methods
+    },
+    suggestions() {
+      const suggestions = [
+        ...this.contextSuggestions,
+        ...this.methodSuggestions,
+      ]
+      return [
+        ...suggestions,
+        this.topLevelSuggestions,
```

Comment:
Hi @poojahpatil90, each time we use the spread (`...`) operator, we end up copying the lists. Could we avoid 1 copy by dropping them directly into the return statement's list? Not sure if that makes sense, but could we reduce this to something like:
```
return [
  ...this.contextSuggestions,
  ...this.methodSuggestions,
  ...this.topLevelSuggestions,
]
```

Reply:
This makes sense!  I did not notice this part, I have optimized this part 👍